posts/2021-04-19-JAXRS-multipart-payloads-21005-beta.adoc

```diff
@@ -0,0 +1,193 @@
+* <<jakarta, Jakarta EE 9 Beta Features>>: a lightweight package that contains only the Jakarta EE 9 features.
+
+This means that you can now try out our in-development Open Liberty features by just adding the relevant coordinates to your build tools.
+
+If you try either package, <<feedback, let us know what you think>>.
+[#allbeta]
+== All Beta Features package
+
+The All Beta Features package includes 
+
+* <<webservices, Build multipart payloads for your JAX-RS client and services>>.
+* <<jee9support, Jakarta EE 9 support for oauth-2.0 and samlWeb-2.0>>s
+
+[#webservices]
+=== Build multipart payloads for your JAX-RS client and services
+
+Often, a RESTful service or client will need to send multiple disparate pieces of data in the same request, for example, uploading a resume/CV with a picture and text for name and address. This is usually done using `multipart/form-data`.
+
+While Liberty currently has APIs to enable users to _receive_ multipart/form-data payloads, it does not have any APIs to enable users to _send_ multipart payloads - until now. With the new `AttachmentBuilder` API, users can now send multipart requests from their JAX-RS clients or send multipart payloads as responses from their JAX-RS resources.
+
+To send a multipart request, you will need to enable the `jaxrs-2.0` or `jaxrs-2.1` feature. Presumably, if you are already using JAX-RS, one of these features will already be enabled. To send a multipart payload, you must send an instance of `List<IAttachment>`.  Each object in that list represents a single attachment part.  Here is an example of creating and sending a multipart request from a JAX-RS client:
+[source, java]
+----
+List<IAttachment> attachments = new ArrayList<>();
+
+attachments.add(AttachmentBuilder.newBuilder("blogPost")
+                                 .inputStream(new FileInputStream("/path/to/yesterdaysBlogPost.xml"))
+                                 .fileName("myRenamedBlogPost.asciidoc")
+                                 .contentType("text/asciidoc")
+                                 .contentId("myBlogPostID")
+                                 .header("X-PriorityLevel", "Medium")
+                                 .build());
+
+attachments.add(AttachmentBuilder.newBuilder("file1")
+                                 .inputStream("some.xml", new FileInputStream("/path/to/myPicture.png"))
+                                 .contentType("image/png")
+                                 .build());
+
+attachments.add(AttachmentBuilder.newBuilder("authorName")
+                                 .inputStream(new ByteArrayInputStream("John Doe".getBytes()))
+                                 .build());
+
+Response response = client.target(BLOG_SITE_URI)
+                          .request()
+                          .post(Entity.entity(attachments, MediaType.MULTIPART_FORM_DATA));
+----
+
```

Comment:
@andymc12 you don't happen to have an image/screenshot of this multipart payload going over the wire (something like the browser network tab, or even wireshark)?  If you did then we could include it here at the end and set it as the image for the entire blog post via `open-graph-image` tag

Reply:
@mbroz2 no, not really.  I could probably come up with something, but it'd be a bit contrived (e.g. sending"Hello World" in multiple parts) because most real-world use cases would be sending files that would be too large to put in a concise image.    I'll see what I can do, but please don't hold up the beta blog for me. :) Thanks.